# ls command


In [8]:
%%bash

# Create a test directory and navigate into it
mkdir -p test_dir/level1
cd test_dir

# Create some files and subdirectories
touch file1 file2 file3
mkdir subdir1 subdir2
cd ..
tree test_dir


mkdir: cannot create directory ‘subdir1’: File exists
mkdir: cannot create directory ‘subdir2’: File exists


test_dir
├── file1
├── file2
├── file3
├── level1
├── subdir1
└── subdir2

4 directories, 3 files


In [21]:
%%bash
cd 
echo "# Default listing of files and directories in the current directory"
ls 
echo ""
echo "# List in long format (-l) showing permissions, owner, group, size, and timestamp"
ls -l
echo ""
echo "# List with human-readable file sizes (-lh) to display sizes in KB, MB, etc."
ls -lh


# Default listing of files and directories in the current directory
rep1
test_dir
text_demo

# List in long format (-l) showing permissions, owner, group, size, and timestamp
total 8
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
drwxr-xr-x 5 kavis kavis 4096 Oct  9 02:44 test_dir
drwxr-xr-x 2 kavis kavis 4096 Oct  8 10:28 text_demo

# List with human-readable file sizes (-lh) to display sizes in KB, MB, etc.
total 8.0K
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
drwxr-xr-x 5 kavis kavis 4.0K Oct  9 02:44 test_dir
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 10:28 text_demo


In [22]:
%%bash
echo "Show hidden files as well (-a), includes files starting with '.'"
ls -a
echo " "

echo "Long format with hidden files (-la), useful to see permissions of hidden files"
ls -la


Show hidden files as well (-a), includes files starting with '.'
.
..
.hidden_directory
.hidden_file.txt
1500x500.jpeg
Linux_shell_setup_windows.ipynb
Linux_shell_w1_a.ipynb
Linux_shell_w1_b.ipynb
Linux_shell_w1_c.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb
demo
random_toss.py
test_dir
 
Long format with hidden files (-la), useful to see permissions of hidden files
total 620
drwxrwxrwx 1 kavis kavis   4096 Oct  9 02:59 .
drwxrwxrwx 1 kavis kavis   4096 Oct  6 11:50 ..
drwxrwxrwx 1 kavis kavis   4096 Oct  8 08:38 .hidden_directory
-rwxrwxrwx 1 kavis kavis      0 Oct  8 08:38 .hidden_file.txt
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis 126332 Oct  8 19:22 Linux_shell_w1_a.ipynb
-rwxrwxrwx 1 kavis kavis  44199 Oct  9 02:55 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis    234 Oct  8 18:55 Linux_shell_w1_c.i

In [11]:

%%bash
echo "List in long format with hidden files and include inode number (-lia)"
ls -lia


List in long format with hidden files and include inode number (-lia)
total 620
 1970324837338733 drwxrwxrwx 1 kavis kavis   4096 Oct  9 02:59 .
 1688849860627019 drwxrwxrwx 1 kavis kavis   4096 Oct  6 11:50 ..
18014398509607323 drwxrwxrwx 1 kavis kavis   4096 Oct  8 08:38 .hidden_directory
23643898043815456 -rwxrwxrwx 1 kavis kavis      0 Oct  8 08:38 .hidden_file.txt
22236523160207455 -rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
 1688849860628098 -rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
38562071809434349 -rwxrwxrwx 1 kavis kavis 126332 Oct  8 19:22 Linux_shell_w1_a.ipynb
 5910974511104067 -rwxrwxrwx 1 kavis kavis  44199 Oct  9 02:55 Linux_shell_w1_b.ipynb
 9007199254921272 -rwxrwxrwx 1 kavis kavis    234 Oct  8 18:55 Linux_shell_w1_c.ipynb
 1970324837338746 -rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
 1688849860628082 -rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
 1688849860628085 -rwxrwxrwx 1

In [23]:
%%bash
echo " "
echo "Show the inode numbers (-i) of files and directories in the current directory"
ls -i


 
Show the inode numbers (-i) of files and directories in the current directory
22236523160207455 1500x500.jpeg
 1688849860628098 Linux_shell_setup_windows.ipynb
38562071809434349 Linux_shell_w1_a.ipynb
 5910974511104067 Linux_shell_w1_b.ipynb
 9007199254921272 Linux_shell_w1_c.ipynb
 1970324837338746 SC_Week0_Unit0.ipynb
 1688849860628082 SC_Week0_Unit1.ipynb
 1688849860628085 SC_Week0_Unit2.ipynb
 1688849860628079 SC_Week0_Unit4.ipynb
 1688849860628086 SC_Week0_Unit5.ipynb
23925373020449842 demo
21673573206783692 random_toss.py
10696049115641040 test_dir


In [12]:
%%bash
echo "List in long format of only directories, showing details of directories themselves (-ld)"
ls -ld */
echo " "

echo "Show inode numbers along with hidden files in a directory listing (-ia)"
ls -ia

List in long format of only directories, showing details of directories themselves (-ld)
drwxrwxrwx 1 kavis kavis 4096 Oct  8 19:06 demo/
drwxrwxrwx 1 kavis kavis 4096 Oct  9 02:59 test_dir/
 
Show inode numbers along with hidden files in a directory listing (-ia)
 1970324837338733 .
 1688849860627019 ..
18014398509607323 .hidden_directory
23643898043815456 .hidden_file.txt
22236523160207455 1500x500.jpeg
 1688849860628098 Linux_shell_setup_windows.ipynb
38562071809434349 Linux_shell_w1_a.ipynb
 5910974511104067 Linux_shell_w1_b.ipynb
 9007199254921272 Linux_shell_w1_c.ipynb
 1970324837338746 SC_Week0_Unit0.ipynb
 1688849860628082 SC_Week0_Unit1.ipynb
 1688849860628085 SC_Week0_Unit2.ipynb
 1688849860628079 SC_Week0_Unit4.ipynb
 1688849860628086 SC_Week0_Unit5.ipynb
23925373020449842 demo
21673573206783692 random_toss.py
10696049115641040 test_dir


In [13]:
%%bash
echo "Show file sizes in blocks (-s)"
ls -s
echo " "


Show file sizes in blocks (-s)
total 620
164 1500x500.jpeg
 24 Linux_shell_setup_windows.ipynb
124 Linux_shell_w1_a.ipynb
 44 Linux_shell_w1_b.ipynb
  0 Linux_shell_w1_c.ipynb
  8 SC_Week0_Unit0.ipynb
228 SC_Week0_Unit1.ipynb
 16 SC_Week0_Unit2.ipynb
  4 SC_Week0_Unit4.ipynb
  8 SC_Week0_Unit5.ipynb
  0 demo
  0 random_toss.py
  0 test_dir
 


In [14]:
%%bash
echo "Recursive listing, show contents of all directories and subdirectories (-R)"
ls -R


Recursive listing, show contents of all directories and subdirectories (-R)
.:
1500x500.jpeg
Linux_shell_setup_windows.ipynb
Linux_shell_w1_a.ipynb
Linux_shell_w1_b.ipynb
Linux_shell_w1_c.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb
demo
random_toss.py
test_dir

./demo:
leve1

./demo/leve1:
level2

./demo/leve1/level2:

./test_dir:
file1
file2
file3
level1
subdir1
subdir2

./test_dir/level1:

./test_dir/subdir1:

./test_dir/subdir2:


In [15]:

%%bash
echo "Recursive listing in long format (-lR)"
ls -lR


Recursive listing in long format (-lR)
.:
total 620
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis 126332 Oct  8 19:22 Linux_shell_w1_a.ipynb
-rwxrwxrwx 1 kavis kavis  44199 Oct  9 02:55 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis    234 Oct  8 18:55 Linux_shell_w1_c.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
drwxrwxrwx 1 kavis kavis   4096 Oct  8 19:06 demo
-rwxrwxrwx 1 kavis kavis    199 Oct  6 19:47 random_toss.py
drwxrwxrwx 1 kavis kavis   4096 Oct  9 02:59 test_dir

./demo:
total 0
drwxrwxrwx 1 kavis kavis 4096 Oct  8 19:06 leve1

./demo/leve1:
total 0
drwxrwxrwx 1 kavis kavis 

In [16]:
%%bash
# Show files sorted by size (-lS) to get largest files first
ls -lS

# List files sorted by modification time, newest first (-lt)
ls -lt

# List files in reverse order by modification time (-ltr)
ls -ltr


total 620
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis 126332 Oct  8 19:22 Linux_shell_w1_a.ipynb
-rwxrwxrwx 1 kavis kavis  44199 Oct  9 02:55 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
drwxrwxrwx 1 kavis kavis   4096 Oct  8 19:06 demo
drwxrwxrwx 1 kavis kavis   4096 Oct  9 02:59 test_dir
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis    234 Oct  8 18:55 Linux_shell_w1_c.ipynb
-rwxrwxrwx 1 kavis kavis    199 Oct  6 19:47 random_toss.py
total 620
drwxrwxrwx 1 kavis kavis   4096 Oct  9 02:59 test_dir
-rwxrwxrwx 1 kavis kavis  44199 Oct  9 02:55 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis

In [17]:
%%bash
# Combine -i and -l to show inode numbers along with detailed view (-il)
ls -il

# Combine -i, -a, and -l to show inode numbers, hidden files, and detailed view (-ial)
ls -ial


total 620
22236523160207455 -rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
 1688849860628098 -rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
38562071809434349 -rwxrwxrwx 1 kavis kavis 126332 Oct  8 19:22 Linux_shell_w1_a.ipynb
 5910974511104067 -rwxrwxrwx 1 kavis kavis  44199 Oct  9 02:55 Linux_shell_w1_b.ipynb
 9007199254921272 -rwxrwxrwx 1 kavis kavis    234 Oct  8 18:55 Linux_shell_w1_c.ipynb
 1970324837338746 -rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
 1688849860628082 -rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
 1688849860628085 -rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
 1688849860628079 -rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
 1688849860628086 -rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
23925373020449842 drwxrwxrwx 1 kavis kavis   4096 Oct  8 19:06 demo
21673573206783692 -rwxrwxrwx 1 kavis kavis    199 Oct  6 19:47 random_to

In [18]:
%%bash
# Using a directory as an argument, list contents of "level1" directory
cd test_dir/level1


In [25]:
%%bash

# Navigate into the "level1" directory and list contents
cd test_dir/level1
pwd
ls -l

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell/test_dir/level1


total 0


In [27]:
%%bash
# Go back to parent directory and list contents recursively with details (-lR)
ls -lR


.:
total 620
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis 126332 Oct  8 19:22 Linux_shell_w1_a.ipynb
-rwxrwxrwx 1 kavis kavis  44199 Oct  9 02:55 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis    234 Oct  8 18:55 Linux_shell_w1_c.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
drwxrwxrwx 1 kavis kavis   4096 Oct  8 19:06 demo
-rwxrwxrwx 1 kavis kavis    199 Oct  6 19:47 random_toss.py
drwxrwxrwx 1 kavis kavis   4096 Oct  9 02:59 test_dir

./demo:
total 0
drwxrwxrwx 1 kavis kavis 4096 Oct  8 19:06 leve1

./demo/leve1:
total 0
drwxrwxrwx 1 kavis kavis 4096 Oct  8 19:06 level2

./demo/leve1/

In [28]:
%%bash
# Use -d to show information only about the directory itself, not its contents
ls -ld test_dir/


drwxrwxrwx 1 kavis kavis 4096 Oct  9 02:59 test_dir/
22236523160207455 1500x500.jpeg
 1688849860628098 Linux_shell_setup_windows.ipynb
38562071809434349 Linux_shell_w1_a.ipynb
 5910974511104067 Linux_shell_w1_b.ipynb
 9007199254921272 Linux_shell_w1_c.ipynb
 1970324837338746 SC_Week0_Unit0.ipynb
 1688849860628082 SC_Week0_Unit1.ipynb
 1688849860628085 SC_Week0_Unit2.ipynb
 1688849860628079 SC_Week0_Unit4.ipynb
 1688849860628086 SC_Week0_Unit5.ipynb
23925373020449842 demo
21673573206783692 random_toss.py
10696049115641040 test_dir
total 620K
-rwxrwxrwx 1 kavis kavis 226K Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis 163K Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis 124K Oct  8 19:22 Linux_shell_w1_a.ipynb
-rwxrwxrwx 1 kavis kavis  44K Oct  9 02:55 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis  23K Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  16K Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis 8.0K Oct  6 19:54 SC_Week0_Unit0.ipy

In [30]:
%%bash
# List all files including hidden ones and inodes, but just at the top-level directory (-iAd)
ls -iAd *

echo " "
# List files with detailed view and human-readable sizes, sort by size (-lhS)
ls -lhS


22236523160207455 1500x500.jpeg
 1688849860628098 Linux_shell_setup_windows.ipynb
38562071809434349 Linux_shell_w1_a.ipynb
 5910974511104067 Linux_shell_w1_b.ipynb
 9007199254921272 Linux_shell_w1_c.ipynb
 1970324837338746 SC_Week0_Unit0.ipynb
 1688849860628082 SC_Week0_Unit1.ipynb
 1688849860628085 SC_Week0_Unit2.ipynb
 1688849860628079 SC_Week0_Unit4.ipynb
 1688849860628086 SC_Week0_Unit5.ipynb


23925373020449842 demo
21673573206783692 random_toss.py
10696049115641040 test_dir
 
total 620K
-rwxrwxrwx 1 kavis kavis 226K Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis 163K Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis 124K Oct  8 19:22 Linux_shell_w1_a.ipynb
-rwxrwxrwx 1 kavis kavis  44K Oct  9 02:55 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis  23K Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  16K Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis 8.0K Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 4.1K Oct  5 14:21 SC_Week0_Unit5.ipynb
drwxrwxrwx 1 kavis kavis 4.0K Oct  8 19:06 demo
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 02:59 test_dir
-rwxrwxrwx 1 kavis kavis 3.5K Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis  234 Oct  8 18:55 Linux_shell_w1_c.ipynb
-rwxrwxrwx 1 kavis kavis  199 Oct  6 19:47 random_toss.py


In [32]:
%%bash
# List directories themselves in detailed view with inode numbers (-ild)
ls -ild */


23925373020449842 drwxrwxrwx 1 kavis kavis 4096 Oct  8 19:06 demo/
10696049115641040 drwxrwxrwx 1 kavis kavis 4096 Oct  9 02:59 test_dir/


In [34]:
%%bash
# Create a test directory structure with nested subdirectories and files
mkdir -p level1/level2
touch file1 file2 level1/file3 level1/level2/file4
mkdir subdir1 subdir2
cd ..
tree Linux_shell


Linux_shell
├── 1500x500.jpeg
├── Linux_shell_setup_windows.ipynb
├── Linux_shell_w1_a.ipynb
├── Linux_shell_w1_b.ipynb
├── Linux_shell_w1_c.ipynb
├── SC_Week0_Unit0.ipynb
├── SC_Week0_Unit1.ipynb
├── SC_Week0_Unit2.ipynb
├── SC_Week0_Unit4.ipynb
├── SC_Week0_Unit5.ipynb
├── demo
│   └── leve1
│       └── level2
├── file1
├── file2
├── level1
│   ├── file3
│   └── level2
│       └── file4
├── random_toss.py
├── subdir1
└── subdir2

8 directories, 15 files


In [36]:
%%bash
# List contents of a specific directory using it as an argument
ls level1


file3
level2


In [ ]:
%%bash
# List contents of multiple directories (subdir1, level1) using directory names as arguments
ls subdir1 level1

# List contents of multiple directories in long format (-l)
ls -l subdir1 level1


In [ ]:
%%bash
# List the contents of level1 recursively (-R), showing all files in subdirectories
ls -R level1

# List detailed view of the level1 directory itself without showing its contents (-ld)
ls -ld level1


In [ ]:
%%bash
# Using -ld to show information about the directory "level1/level2" without listing its contents
ls -ld level1/level2


In [ ]:
%%bash
# List with detailed view and human-readable sizes (-lh) for multiple directories
ls -lh level1 subdir1


In [ ]:

# Recursive listing for multiple directories (-lR)
ls -lR level1 subdir2



In [ ]:
%%bash
# Combine options to show all directories and subdirectories recursively with human-readable sizes (-alhR)
ls -alhR level1 subdir1 subdir2


In [ ]:

%%bash
# Use a combination of -i (inode numbers) and directory arguments
ls -i level1 level1/level2



In [ ]:
%%bash
# List inode numbers and long format with human-readable sizes for multiple directories (-ilh)
ls -ilh level1 subdir1 subdir2


In [ ]:

# Clean up: remove the created test directory
rm -r test_dir


In [ ]:
'''
%%bash

cd
pwd

# Create a sample directory structure
mkdir -p sample_dir/subdir
cd sample_dir
touch file1 file2 .hidden_file subdir/file3

# Short form of options:
ls -a          # Show all files, including hidden files
ls -l          # Show long format listing
ls -la         # Combine -l and -a to show hidden files in long format
ls -lh         # Long format with human-readable file sizes
ls -R          # Recursively list all directories and their contents
ls -d */       # Show directory names only
ls -i          # Show inode numbers
ls -r          # Reverse the order of the listing
ls -S          # Sort by file size
ls -t          # Sort by modification time

# Long form of options:
ls --all           # Equivalent to -a, show all files including hidden ones
ls --long          # Equivalent to -l, show long format listing
ls --all --long    # Equivalent to -la, show hidden files in long format
ls --human-readable --long  # Long format with human-readable sizes
ls --recursive     # Equivalent to -R, show directories and their contents recursively
ls --directory */  # Equivalent to -d, show directory names only
ls --inode         # Equivalent to -i, show inode numbers
ls --reverse       # Equivalent to -r, show files in reverse order
ls --sort=size     # Equivalent to -S, sort by file size
ls --time          # Equivalent to -t, sort by modification time

# Clean up: remove the created test directory
cd ..
rm -r sample_dir
'''


Key Points to Remember:
Short Form: Typically a single character preceded by a hyphen (-), e.g., -l, -a, -R.

Long Form: A descriptive word preceded by two hyphens (--), e.g., --all, --recursive, --long.

Short and long forms can be used interchangeably. For example, -l is equivalent to --long.

Multiple short options can be combined (e.g., ls -laR), but long options must be written separately (e.g., ls --long --all --recursive).

Use man ls or ls --help to see all available short and long forms for ls.

Short Form vs. Long Form of ls Options

Short Form	Long Form	Description
-a	          --all	    Show all files, including hidden files (those starting with .).

-A	       --almost-all	Show all files except . (current directory) and .. (parent).

-d	     --directory	List directory names only, not their contents.

-h	--human-readable	Show sizes in human-readable format (KB, MB, etc.).

-l	          --long	Use a long listing format to show detailed information.
-R	     --recursive	List subdirectories recursively.

-S	     --sort=size	Sort files by size (largest first).

-t	          --time	Sort by modification time (newest first).

-r	--reverse	Reverse the order of sorting.

-i	--inode	Show inode numbers.

-s	--size	Display file sizes in blocks.

-u	--access-time	Sort by last access time instead of modification time.

-c	--time=ctime	Sort by change time (status change time).

-G	--no-group	Do not show group information in the long format.

-X	--sort=extension	Sort alphabetically by file extension.

-1	--format=single-column	List one file per line.

This sequence includes a broad range of options for ls:

-l: Long format.
-h: Human-readable sizes.
-a: Include hidden files.
-i: Show inode numbers.
-d: Show directory information only.
-R: Recursive listing.
-S: Sort by size.
-t: Sort by modification time.
-r: Reverse the order.
-ia: Include inode numbers and hidden files.

Explanation of ls with Directories as Arguments:


ls <directory_name>: Lists the contents of the specified directory, not the current working directory.
Example: ls level1 lists files inside level1 directory.


ls <dir1> <dir2>: Lists contents of multiple directories.
Example: ls level1 subdir1 lists files in both level1 and subdir1.


ls -ld <directory>: Lists information about the directory itself, not its contents.
Example: ls -ld level1 shows details (like permissions, owner, etc.) of the directory level1 itself.


Recursive Listing (ls -R <directory>): Displays all files and subdirectories within the specified directory recursively.
Example: ls -R level1 shows all files in level1 and its subdirectories.


Combine Options with Directory Arguments:
ls -alhR level1 subdir1: Shows all hidden files in a long format, human-readable sizes, and recursively lists contents of level1 and subdir1.

In [38]:
%%bash
pwd
cd /mnt/c
ls -l
clear

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell


ls: DumpStack.log: Permission denied
ls: cannot access 'DumpStack.log.tmp': Permission denied
ls: cannot access 'hiberfil.sys': Permission denied
ls: cannot access 'pagefile.sys': Permission denied
ls: PerfLogs: Permission denied
ls: Recovery: Permission denied
ls: cannot access 'swapfile.sys': Permission denied
ls: 'System Volume Information': Permission denied


total 16
drwxrwxrwx 1 kavis kavis  4096 Oct 20  2022 $Recycle.Bin
-r-xr-xr-x 1 kavis kavis    40 Oct 20  2022 2CA0AE717CAA
lrwxrwxrwx 1 kavis kavis    12 Oct 10  2022 Documents and Settings -> /mnt/c/Users
drwxrwxrwx 1 kavis kavis  4096 Jul 30  2022 Drivers
---------- 1 kavis kavis 12288 Jul 12 03:44 DumpStack.log
-????????? ? ?     ?         ?            ? DumpStack.log.tmp
drwxrwxrwx 1 kavis kavis  4096 May  4 08:13 OneDriveTemp
d--x--x--x 1 kavis kavis  4096 May  7  2022 PerfLogs
dr-xr-xr-x 1 kavis kavis  4096 Oct  6 03:43 Program Files
dr-xr-xr-x 1 kavis kavis  4096 Jun 11 03:17 Program Files (x86)
drwxrwxrwx 1 kavis kavis  4096 Sep 19 16:32 ProgramData
d--x--x--x 1 kavis kavis  4096 Aug 10  2023 Recovery
-r-xr-xr-x 1 kavis kavis  1505 Dec 19  2023 Setup.log
d--x--x--x 1 kavis kavis  4096 Oct  7 17:23 System Volume Information
drwxrwxrwx 1 kavis kavis  4096 Dec  3  2023 Tenorshare
dr-xr-xr-x 1 kavis kavis  4096 Jul 10 09:50 Users
dr-xr-xr-x 1 kavis kavis  4096 Oct  5 11:03 Windows



#### Structure of Each Line given by ls -l:
Structure will change slightly depending on VS code is run in wsl or not

1. **Permissions**: (`drwxr-xr-x`)
   - This shows the file permissions for the owner, group, and others.
   - first letter -  `d` indicates a **directory** (if it were a file, it would be `-`).
                     'l' indicates symbolic link (shortcut)
                     '-' indicates file 
                     'c' character device (terminal)
                     'b' block devices (hard discs)
   - next three for owner - `rwx` means **read, write, and execute** permissions for the owner.
   - next three for group - `r-x` means **read and execute** permissions for the group.
   - next three for others- `r-x` means **read and execute** permissions for others.

   rwx (7)
   rw- (6)
   r-x (5)
   r-- (4)
   --x (1)

2. **Links**: (`1`)
   - The number of links or directories inside this directory. For a directory, this number (1) represents how many subdirectories, parent directories, or other references link to it. A hard link is like an alias or shortcut that points to the same file or directory in the filesystem. 

3. **Owner**: (`kavis`)
   - The username of the owner of the file or directory.

4. **Group**: (`197609`)
   - The group ID associated with the file.

5. **Size**: (`0`, `40`, `12288`, etc.)
   - The size of the file in bytes. For directories, it’s usually `0` since it doesn’t directly reflect the size of contents.

6. **Timestamp**: (`Oct 6 14:21`)
   - The date and time when the file or directory was last modified.

7. **File/Directory Name**: (`$Recycle.Bin`, `Documents and Settings`, etc.)
   - The name of the file or directory.

#### Listing Explanation:

**`Documents and Settings -> /c/Users`**:
   - This is a **symbolic link** (shortcut) that points to `/c/Users`. It means that trying to access `Documents and Settings` will actually redirect you to the `/c/Users` directory.

**`Drivers`**:
   - Likely a folder containing drivers for the system. Drivers are software that help the operating system communicate with hardware devices.

**`Program Files`** and **`Program Files (x86)`**:
   - Standard directories in Windows for installing 64-bit and 32-bit applications, respectively.

**`ProgramData`**:
   - A directory used to store application data that is not specific to a user. It’s shared among all users of the system.

**`Users`**:
    - This is where user profiles are stored on Windows, such as `Desktop`, `Documents`, `Downloads`, and so on for each user.

**`Windows`**:
    - The main directory containing the Windows operating system files.

**`hiberfil.sys`**:
    - A **system file** used by Windows to store the system state when it goes into **hibernation**. This file is often large and contains everything stored in memory (RAM).

In [ ]:
%%bash
pwd
cd /
ls -im  #ls -i shows the inode number associated with each file or directory, which helps identify them uniquely
ls -l

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
12 bin, 8193 bin.usr-is-merged, 8194 boot, 1 dev, 16386 etc, 24577 home, 5 init,
1395 lib, 24578 lib.usr-is-merged, 1396 lib64, 11 lost+found, 32769 media,
32770 mnt, 40961 opt, 1 proc, 49153 root, 1 run, 1424 sbin,
57345 sbin.usr-is-merged, 57346 snap, 65537 srv, 1 sys, 73729 tmp, 73730 usr,
33554433 var
total 2284
lrwxrwxrwx   1 root root       7 Apr 22 13:08 bin -> usr/bin
drwxr-xr-x   2 root root    4096 Feb 26  2024 bin.usr-is-merged
drwxr-xr-x   2 root root    4096 Apr 22 13:08 boot
drwxr-xr-x  16 root root    3560 Oct  8 08:09 dev
drwxr-xr-x  92 root root    4096 Oct  8 08:09 etc
drwxr-xr-x   3 root root    4096 Oct  5 12:54 home
-rwxrwxrwx   1 root root 2260248 Sep 27 00:55 init
lrwxrwxrwx   1 root root       7 Apr 22 13:08 lib -> usr/lib
drwxr-xr-x   2 root root    4096 Apr  8  2024 lib.usr-is-merged
lrwxrwxrwx   1 root root       9 Apr 22 13:08 lib64 -> usr/lib64
drwx------   2 root root   16384 Oct  5 12:49 los

/bin -       essential command binaries - files in machine language for processor

/usr/bin -   User commands

/usr/lib -   Libraries